crates/ide-assists/src/handlers/generate_fn_type_alias.rs

```diff
@@ -0,0 +1,442 @@
+use either::Either;
+use hir::HirDisplay;
+use ide_db::assists::{AssistId, AssistKind, GroupLabel};
+use syntax::{
+    ast::{self, edit::IndentLevel, make, HasGenericParams, HasName},
+    syntax_editor, AstNode,
+};
+
+use crate::{AssistContext, Assists};
+
+// Assist: generate_fn_type_alias_named
+//
+// Generate a type alias for the function with named parameters.
+//
+// ```
+// unsafe fn fo$0o(n: i32) -> i32 { 42i32 }
+// ```
+// ->
+// ```
+// type ${0:FooFn} = unsafe fn(n: i32) -> i32;
+//
+// unsafe fn foo(n: i32) -> i32 { 42i32 }
+// ```
+
+// Assist: generate_fn_type_alias_unnamed
+//
+// Generate a type alias for the function with unnamed parameters.
+//
+// ```
+// unsafe fn fo$0o(n: i32) -> i32 { 42i32 }
+// ```
+// ->
+// ```
+// type ${0:FooFn} = unsafe fn(i32) -> i32;
+//
+// unsafe fn foo(n: i32) -> i32 { 42i32 }
+// ```
+
+pub(crate) fn generate_fn_type_alias(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    let name = ctx.find_node_at_offset::<ast::Name>()?;
+    let func = &name.syntax().parent()?;
+    let item = func.ancestors().find_map(ast::Item::cast)?;
+    let assoc_owner =
+        item.syntax().ancestors().nth(2).and_then(Either::<ast::Trait, ast::Impl>::cast);
+    let node = assoc_owner.as_ref().map_or_else(
+        || item.syntax(),
+        |impl_| impl_.as_ref().either(AstNode::syntax, AstNode::syntax),
+    );
+    let func_node = ast::Fn::cast(func.clone())?;
+    let param_list = func_node.param_list()?;
+
+    for style in ParamStyle::ALL {
+        let generic_params = func_node.generic_param_list();
+        let module = match ctx.sema.scope(node) {
+            Some(scope) => scope.module(),
+            None => continue,
+        };
+
+        acc.add_group(
+            &GroupLabel("Generate a type alias for function...".into()),
+            style.assist_id(),
+            style.label(),
+            func_node.syntax().text_range(),
+            |builder| {
+                let mut edit = builder.make_editor(func);
+
+                let alias_name = format!("{}Fn", stdx::to_camel_case(&name.to_string()));
+
+                let fn_abi = match func_node.abi() {
+                    Some(abi) => format!("{} ", abi),
+                    None => "".into(),
+                };
+
+                let fn_unsafe = if func_node.unsafe_token().is_some() { "unsafe " } else { "" };
+
+                let fn_qualifiers = format!("{fn_unsafe}{fn_abi}");
+
+                let fn_type = return_type(&func_node);
+
+                let mut fn_params_vec = Vec::new();
+
+                if let Some(self_param) = param_list.self_param() {
+                    if let Some(local) = ctx.sema.to_def(&self_param) {
+                        let ty = local.ty(ctx.db());
+                        if let Ok(s) = ty.display_source_code(ctx.db(), module.into(), false) {
+                            fn_params_vec.push(s)
+                        }
+                    }
+                }
```

Comment:
Seeing this, it might be good to generally work with the hir version of a function instead of picking apart its syntax node and reconstructing it? That is wrok with the result of `ctx.sema.to_def(&func_node)`